packages/hdwallet-provider/src/index.ts

```diff
@@ -28,6 +28,127 @@ import { Callback, JsonRPCResponse } from "web3/providers";
 // See issue #65 for more
 const singletonNonceSubProvider = new NonceSubProvider();
 
+export interface MnemonicOptions {
+  phrase: string;
+  password?: string;
+}
+
+export interface Mnemonic {
+  phrase: string;
+  password: string;
+}
+
+export interface PrivateKeysOptions {
+  privateKeys: string | string[];
+}
+
+export interface CommonConstructorOptions {
+  url: string;
+  addressIndex?: number;
+  numberOfAddresses?: number;
+  sharedNonce?: boolean;
+  derivationPath?: string;
+}
+
+export type ConstructorOptions =
+  | (CommonConstructorOptions & MnemonicOptions)
+  | (CommonConstructorOptions & PrivateKeysOptions);
+
+const OPTIONS_DEFAULTS = {
+  ADDRESS_INDEX: 0,
+  NUMBER_OF_ADDRESSES: 10,
+  SHARED_NONCE: true,
+  DERIVATION_PATH: `m/44'/60'/0'/0/`
+};
+
+const getOptions = (
+  mnemonicOrPrivateKeysOrOptions: string | string[] | ConstructorOptions,
+  options: any[]
+): {
+  url: string;
+  addressIndex: number;
+  numberOfAddresses: number;
+  sharedNonce: boolean;
+  derivationPath: string;
+} => {
+  if (options.length) {
+    const [
+      url,
+      addressIndex = OPTIONS_DEFAULTS.ADDRESS_INDEX,
+      numberOfAddresses = OPTIONS_DEFAULTS.NUMBER_OF_ADDRESSES,
+      sharedNonce = OPTIONS_DEFAULTS.SHARED_NONCE,
+      derivationPath = OPTIONS_DEFAULTS.DERIVATION_PATH
+    ] = options;
+
+    return {
+      addressIndex,
+      numberOfAddresses,
+      sharedNonce,
+      derivationPath,
+      url
+    };
+  }
+
+  const {
+    url,
+    addressIndex = OPTIONS_DEFAULTS.ADDRESS_INDEX,
+    numberOfAddresses = OPTIONS_DEFAULTS.NUMBER_OF_ADDRESSES,
+    sharedNonce = OPTIONS_DEFAULTS.SHARED_NONCE,
+    derivationPath = OPTIONS_DEFAULTS.DERIVATION_PATH
+  } = mnemonicOrPrivateKeysOrOptions as ConstructorOptions;
+
+  return { addressIndex, numberOfAddresses, sharedNonce, derivationPath, url };
+};
+
+const getMnemonic = (
+  mnemonicOrPrivateKeysOrOptions: string | string[] | ConstructorOptions
+): { phrase: string | null; password: string } => {
+  const phrase =
+    typeof mnemonicOrPrivateKeysOrOptions === "string" &&
+    mnemonicOrPrivateKeysOrOptions.includes(" ")
```

Comment:
I wonder if there is a better way to do this. I'm going to just throw out an idea here, what if we `split` on " " and test the length? That would make this more lenient in the case where they have a trailing or leading space on a single private key. We could `trim` and it would be acceptable. But perhaps for something like this we should require that they input their options precisely or else fail. What do you think?

Reply:
Whoops, I merged @gnidan's code into my branch and lost this comment. Haha, the only reason I had this code is that I didn't want to change the initial checks. 

I agree that splitting the string into an actual word list and checking the length might be a good idea. From skimming [the original bip39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#generating-the-mnemonic), it appears that any multiple of 3 between 12 and 24 words is valid. 

Let me know if you want me to change what's there now.

Reply:
Missed this comment in my re-review.

Not a comment about the implementation of that check, but see [my comment below](https://github.com/trufflesuite/truffle/pull/3209#discussion_r465457068) for an idea on where this check should live (spoiler: I think it should live inside `isMnemonicPhrase`)

Reply:
Comment about implementation of that check: sounds like something that we should open an issue for instead of putting into this PR